qstack/fields/dori.py

```diff
@@ -0,0 +1,399 @@
+    """
+
+    def grad_num(func, grid, d, **kwargs):
+        g = np.zeros_like(grid)
+        for j in range(3):
+            u = np.eye(1, 3, j)  # unit vector || jth dimension
+            e1  = func(grid+d*u, **kwargs)
+            e2  = func(grid-d*u, **kwargs)
+            e11 = func(grid+2*d*u, **kwargs)
+            e22 = func(grid-2*d*u, **kwargs)
+            g[:,j] = (8.0*e1-8.0*e2 + e22-e11) / (12.0*d)
+        return g
+
+    def compute_k2(coords, mol=None, dm=None, c=None):
+        rho, drho_dr = compute_rho(mol, coords, dm=dm, c=c, deriv=1)
+        k = drho_dr / rho
+        return np.einsum('xi,xi->i', k, k)
+
+    rho = compute_rho(mol, coords, dm=dm, c=c, deriv=0)
+
+    good_idx = np.where(rho>=eps)[0]
+    dori = np.zeros_like(rho)
+    if len(good_idx):
+        k2 = compute_k2(coords[good_idx], mol=mol, dm=dm, c=c)
+        dk2_dr = grad_num(compute_k2, coords[good_idx], d=dx, mol=mol, dm=dm, c=c)
+        dk2_dr_square = np.einsum('ix,ix->i', dk2_dr, dk2_dr)
+        theta = dk2_dr_square / k2**3
+        dori[good_idx] = theta / (theta + 1.0)
+    return dori, rho
+
+
+def dori_on_grid(mol, coords, dm=None, c=None, eps=1e-4, alg='analytical', mem=1, dx=1e-4, progress=False):
+    """Wrapper to compute DORI on a given grid
+
+    Args:
+        mol (pyscf Mole): pyscf Mole object.
+        coords : 2D array of (ngrids,3)
+            Grid coordinates (in Bohr)
+    Kwargs:
+        dm : 2D array of (nao,nao)
+            Density matrix (confilicts with c)
+        c : 1D array of (nao)
+            Density fitting coefficients (confilicts with dm)
+        eps : float
+            density threshold for DORI
+        alg : str
+            [a]nalytical or [n]umerical computation
+        dx : float
+            Step (in Bohr) to take the numerical derivatives
+        mem : float
+            max. memory (GiB) that can be allocated to compute
+            the AO and their derivatives
+        progress : bool
+            if print a progress bar
+
+    Returns:
+        1D array of (ngrids) --- computed DORI
+        1D array of (ngrids) --- electron density
+        1D array of (ngrids) --- electron density * sgn(second eigenvalue of d^2rho/dr^2)
+                                 if density>=eps else 0 (only with alg='analytical').
+    """
+
+    max_size = int(mem * 2**30)  # mem * 1 GiB
+    point_size = 10 * mol.nao * np.float64().itemsize  # memory needed for 1 grid point
+    dgrid = max_size // point_size
+    grid_chunks = range(0, len(coords), dgrid)
+    if progress:
+        grid_chunks = tqdm(grid_chunks)
+
+    rho = np.zeros(len(coords))
+
+    if alg in [*accumulate('analytical')]:
+        drho_dr = np.zeros((3, len(coords)))
+        d2rho_dr2 = np.zeros((3, 3, len(coords)))
+        for i in grid_chunks:
+            s = np.s_[i:i+dgrid]
+            rho_i, drho_dr_i, d2rho_dr2_i = compute_rho(mol, coords[s], dm=dm, c=c, eps=eps)
+            # Yes the data is copied around too much (three times).
```

Comment:
what do you mean 3 times? I only see one here

Reply:
inside of `compute_rho()` as well

Reply:
I see.
by "We tried to make compute_rho() write directly to these but didn't gain a lot." a few lines below, did you mean `rho[s],drho_dr[:,s],d2rho_dr2[:,:,s]= compute_rho(...)` or `= compute_rho(..., out=(rho[s],drho_dr[:,s],d2rho_dr2[:,:,s]))`?

Reply:
the last one! with passing the slice object as well to bypass `drho_dr[:,good_idx] = ret[1]`

Reply:
I didn't think of the first one so used it to shorten the code